File: src/lib/studio/routingGraph.ts
// Cycle detection for the mixer routing graph.
//
// The mixer forms a DAG: tracks output to buses, buses output to buses
// or master, sends read from tracks (or buses, later) and inject into
// buses. If a user ever routes bus A → bus B → bus A, the audio graph
// tries to feed itself and either goes silent or feeds back — either
// way the engine should refuse the edit and surface a clean error to
// the UI instead of half-committing.
//
// Deliberately Tone-free / native-free — same helper used by the web
// engine (engine.ts) AND ported verbatim to Swift (Phase 4). Its
// contract is entirely on the adjacency graph.
//
// Master is treated as a terminal sink: edges INTO master are always
// legal, edges OUT of master don't exist by construction (master has
// no `output` field). If a caller passes an edge whose source is
// MASTER_BUS_ID, that's ignored on principle — master doesn't route.

import { MASTER_BUS_ID } from './session';

export type RoutingEdge = {
  /** Bus id (or a track id when we introduce sends in Phase 4). */
  from: string;
  /** Bus id — MASTER_BUS_ID or a user bus id. */
  to: string;
};

export type CycleResult =
  | { ok: true }
  | { ok: false; cycle: string[] };

/** DFS with white/gray/black coloring. Returns the offending cycle
 * path (from → ... → from) on failure so the caller can surface it
 * verbatim to the user. */
export function findRoutingCycle(edges: RoutingEdge[]): CycleResult {
  const adj = new Map<string, string[]>();
  for (const e of edges) {
    if (e.from === MASTER_BUS_ID) continue; // master has no outgoing edges
    const list = adj.get(e.from) ?? [];
    list.push(e.to);
    adj.set(e.from, list);
  }

  const WHITE = 0, GRAY = 1, BLACK = 2;
  const color = new Map<string, number>();
  const parent = new Map<string, string>();

  const nodes = new Set<string>();
  for (const e of edges) { nodes.add(e.from); nodes.add(e.to); }

  for (const start of nodes) {
    if ((color.get(start) ?? WHITE) !== WHITE) continue;
    // Iterative DFS to avoid stack overflow on pathological graphs.
    const stack: Array<{ node: string; iter: number }> = [{ node: start, iter: 0 }];
    color.set(start, GRAY);
    while (stack.length > 0) {
      const top = stack[stack.length - 1];
      const neighbors = adj.get(top.node) ?? [];
      if (top.iter >= neighbors.length) {
        color.set(top.node, BLACK);
        stack.pop();
        continue;
      }
      const next = neighbors[top.iter++];
      const c = color.get(next) ?? WHITE;
      if (c === GRAY) {
        // Reconstruct cycle from parent chain: walk back from `top.node`
        // until we return to `next`, then append `next` to close it.
        const cycle: string[] = [next];
        let cur: string | undefined = top.node;
        while (cur !== undefined && cur !== next) {
          cycle.push(cur);
          cur = parent.get(cur);
        }
        cycle.push(next);
        cycle.reverse();
        return { ok: false, cycle };
      }
      if (c === WHITE) {
        parent.set(next, top.node);
        color.set(next, GRAY);
        stack.push({ node: next, iter: 0 });
      }
    }
  }

  return { ok: true };
}

/** Format a cycle for a user-facing error message. */
export function formatCycle(cycle: string[]): string {
  return cycle.join(' → ');
}

File: ios/App/App/Studio/RoutingGraph.swift
// RoutingGraph — Swift mirror of src/lib/studio/routingGraph.ts.
//
// Same algorithm, same contract, ported verbatim so the iOS and web
// engines reject the same cyclic bus edits and format cycles the
// same way in error messages.
//
// This file is INTENTIONALLY dependency-free (no Foundation-heavy
// types beyond what the standard library exports) so it can be
// standalone-verified with `swiftc -parse` without pulling the whole
// audio-engine graph.

/// One directed edge in the bus routing graph: `from` (a bus id or,
/// once sends land, a track id) points at `to` (a bus id, potentially
/// MASTER_BUS_ID).
public struct RoutingEdge: Equatable, Sendable {
    public let from: String
    public let to: String

    public init(from: String, to: String) {
        self.from = from
        self.to = to
    }
}

/// Result of a cycle check. On failure `cycle` holds the offending
/// path with the start node repeated at the end so it can be
/// formatted as `a → b → c → a` for the user.
public enum CycleResult: Equatable, Sendable {
    case ok
    case cycle([String])

    public var isOk: Bool {
        if case .ok = self { return true }
        return false
    }
}

/// DFS with white/gray/black coloring. Master is treated as a
/// terminal sink — outbound edges from master are ignored on
/// principle (master doesn't route anywhere).
///
/// - Parameter edges: the current + proposed edges in the graph.
/// - Returns: `.ok` when the graph is a DAG, or `.cycle([...])`
///   with the offending path.
public func findRoutingCycle(_ edges: [RoutingEdge]) -> CycleResult {
    // Adjacency map — drop outbound edges from master by convention.
    var adjacency: [String: [String]] = [:]
    for edge in edges where edge.from != Studio.masterBusId {
        adjacency[edge.from, default: []].append(edge.to)
    }

    // Node set (for iteration order).
    var nodes = Set<String>()
    for edge in edges {
        nodes.insert(edge.from)
        nodes.insert(edge.to)
    }

    let white = 0, gray = 1, black = 2
    var color: [String: Int] = [:]
    var parent: [String: String] = [:]

    for start in nodes {
        if (color[start] ?? white) != white { continue }
        // Iterative DFS with per-node neighbor-iteration cursor —
        // matches the JS implementation to avoid stack overflow on
        // pathological graphs.
        var stack: [(node: String, iter: Int)] = [(start, 0)]
        color[start] = gray
        while let top = stack.last {
            let neighbors = adjacency[top.node] ?? []
            if top.iter >= neighbors.count {
                color[top.node] = black
                stack.removeLast()
                continue
            }
            stack[stack.count - 1].iter += 1
            let next = neighbors[top.iter]
            let c = color[next] ?? white
            if c == gray {
                // Reconstruct cycle: walk back from `top.node` via
                // parent pointers until we return to `next`, then
                // append `next` to close it.
                var cycle: [String] = [next]
                var cur: String? = top.node
                while let node = cur, node != next {
                    cycle.append(node)
                    cur = parent[node]
                }
                cycle.append(next)
                return .cycle(cycle.reversed())
            }
            if c == white {
                parent[next] = top.node
                color[next] = gray
                stack.append((next, 0))
            }
        }
    }

    return .ok
}

/// Substitute the source's OLD outbound edge with the proposed new
/// one, then run findRoutingCycle. Callers use this to pre-check
/// a bus-output rewire before committing to it.
public func wouldEditCycle(
    existingEdges: [RoutingEdge],
    edit: RoutingEdge,
) -> CycleResult {
    let merged = existingEdges.filter { $0.from != edit.from } + [edit]
    return findRoutingCycle(merged)
}

/// Format a cycle for a user-facing error message. Matches the JS
/// `formatCycle` output byte-for-byte.
public func formatCycle(_ cycle: [String]) -> String {
    return cycle.joined(separator: " → ")
}
